pkg/envoy/xds/cache.go

```diff
@@ -0,0 +1,219 @@
+)
+
+// Cache is a key-value container which allows atomically updating entries and
+// incrementing a version number and notifying observers if the cache is actually
+// modified.
+// This cache implementation ignores the proxy node identifiers, i.e. the same
+// resources are available under the same names to all clients..
+type Cache struct {
+	BaseObservableResourceSet
+
+	// locker is the locker to synchronize updates to resources and version.
+	locker lock.RWMutex
+
+	// resources is the map of cached resource name to resource entry.
+	resources map[cacheKey]cacheValue
+
+	// version is the current version of the resources in the cache.
+	version uint64
+}
+
+// cacheKey uniquely identifies a resource.
+type cacheKey struct {
+	// typeURL is the URL that uniquely identifies the resource's type.
+	typeURL string
+
+	// resourceName is the name of the resource, unique among all the resources
+	// of this type.
+	resourceName string
+}
+
+// cacheValue is a cached resource.
+type cacheValue struct {
+	// resource is the resource in this cache entry.
+	resource proto.Message
+
+	// lastModifiedVersion is the version when this resource entry was last
+	// modified.
+	lastModifiedVersion uint64
+}
+
+// NewCache creates a new, empty cache with 0 as its current version.
+func NewCache() *Cache {
+	return &Cache{
+		resources: make(map[cacheKey]cacheValue),
+		version:   0,
+	}
+}
+
+// Tx inserts/updates a set of resources, then deletes a set of resources, then
+// increases the cache's version number atomically if the cache is actually
+// changed.
+func (c *Cache) Tx(typeURL string, upsertedResources map[string]proto.Message, deletedNames []string) {
+	c.locker.Lock()
+	defer c.locker.Unlock()
+
+	cacheIsUpdated := false
+	newVersion := c.version + 1
+
+	cacheLog := log.WithFields(logrus.Fields{
+		logfields.XDSTypeURL:     typeURL,
+		logfields.XDSVersionInfo: newVersion,
+	})
+
+	cacheLog.Debugf("preparing new cache transaction: upserting %d entries, deleting %d entries",
+		len(upsertedResources), len(deletedNames))
+
+	k := cacheKey{
+		typeURL: typeURL,
+	}
+
+	v := cacheValue{
+		lastModifiedVersion: newVersion,
+	}
+
+	for name, value := range upsertedResources {
+		k.resourceName = name
+		oldV, found := c.resources[k]
+		// If the value is unchanged, don't update the entry, to preserve its
+		// lastModifiedVersion. This allows minimizing the frequency of
+		// responses in GetResources.
+		// Calling proto.Message.String is not very cheap, but we assume that
+		// the reduced churn between the clients and the server is worth it.
+		if !found || oldV.resource.String() != value.String() {
+			if found {
+				cacheLog.WithField(logfields.XDSResourceName, name).Debug("updating resource in cache")
+			} else {
+				cacheLog.WithField(logfields.XDSResourceName, name).Debug("inserting resource into cache")
+			}
+			cacheIsUpdated = true
+			v.resource = value
+			c.resources[k] = v
+		}
+	}
+
+	for _, name := range deletedNames {
```

Comment:
I didn't want to make this method too complicated, because I don't think this method needs to be used directly. AFAIK, only `Upsert` and `Delete` are needed for now. Defining `Tx` was a way to factorize the implementations of `Upsert` and `Delete`. I'll just make `Tx` private.

And I don't think this is the final API for the `Cache` type. I'm all for improving it. But I think we'll know more what the needs are when we start using in the rest of the Cilium agent.

Reply:
Fair enough. I had some other places in mind where it might come in handy, as it's pretty generic and handy to distribute events :) Would you expect it to live in this package once we do use it elsewhere or end up in it's own package somewhere?

Reply:
Hmm that would be doable, but we’d have to implement an adapter to stop implementing `GetResources` directly.

Reply:
That was a longer term thought. I didn't intend for you to change anything in this PR!